themes/default/content/docs/intro/concepts/resources/components.md

```diff
@@ -0,0 +1,287 @@
+
+This example demonstrates both the naming convention and how to designate the component resource as the parent:
+
+{{< chooser language "javascript,typescript,python,go,csharp" >}}
+
+{{% choosable language javascript %}}
+
+```javascript
+let bucket = new aws.s3.Bucket(`${name}-bucket`,
+    {/*...*/}, { parent: this });
+```
+
+{{% /choosable %}}
+{{% choosable language typescript %}}
+
+```typescript
+let bucket = new aws.s3.Bucket(`${name}-bucket`,
+    {/*...*/}, { parent: this });
+```
+
+{{% /choosable %}}
+{{% choosable language python %}}
+
+```python
+bucket = s3.Bucket(f"{name}-bucket",
+    opts=pulumi.ResourceOptions(parent=self))
+```
+
+{{% /choosable %}}
+{{% choosable language go %}}
+
+```go
+bucket, err := s3.NewBucket(ctx, fmt.Sprintf("%s-bucket", name),
+    &s3.BucketArgs{ /*...*/ }, pulumi.Parent(myComponent))
+```
+
+{{% /choosable %}}
+{{% choosable language csharp %}}
+
+```csharp
+var bucket = new Aws.S3.Bucket($"{name}-bucket",
+    new Aws.S3.BucketArgs(/*...*/), new CustomResourceOptions { Parent = this });
+```
+
+{{% /choosable %}}
+
+{{< /chooser >}}
+
+## Registering Component Outputs
+
+Component resources can define their own output properties by using register_outputs . The Pulumi engine uses this information to display the logical outputs of the component resource and any changes to those outputs will be shown during an update.
+
+For example, this code registers an S3 bucket’s computed domain name, which won’t be known until the bucket is created:
+
+{{< chooser language "javascript,typescript,python,go,csharp" >}}
+
+{{% choosable language javascript %}}
+
+```javascript
+this.registerOutputs({
+    bucketDnsName: bucket.bucketDomainName,
+})
+```
+
+{{% /choosable %}}
+{{% choosable language typescript %}}
+
+```typescript
+this.registerOutputs({
+    bucketDnsName: bucket.bucketDomainName,
+})
+```
+
+{{% /choosable %}}
+{{% choosable language python %}}
+
+```python
+self.register_outputs({
+    "bucketDnsName": bucket.bucketDomainName
+})
+```
+
+{{% /choosable %}}
+{{% choosable language go %}}
+
+```go
+ctx.RegisterResourceOutputs(myComponent, pulumi.Map{
+    "bucketDnsName": bucket.BucketDomainName,
+})
+```
+
+{{% /choosable %}}
+{{% choosable language csharp %}}
+
+```csharp
+this.RegisterOutputs(new Dictionary<string, object>
+{
+    { "bucketDnsName", bucket.BucketDomainName }
+});
+```
+
+{{% /choosable %}}
+
+{{< /chooser >}}
+
+The call to `registerOutputs` typically happens at the very end of the component resource’s constructor.
+
+The call to `registerOutputs` also tells Pulumi that the resource is done registering children and should be considered fully constructed, so—although it’s not enforced—the best practice is to call it in all components even if no outputs need to be registered.
```

Comment:
This is an interesting bit. So registerOutputs call is truly optional, we support not calling it?

Reply:
Yes.  The only affect it has today is to update the `updating...` display in the CLI to `updated` on the component itself.  

As we've discussed recently, we should make a breaking change to require this, as an indication of when the component construction is complete so that dependencies on the component are processed at that point in time (and can see "all the children".  This is tracked in https://github.com/pulumi/pulumi/issues/2653, https://github.com/pulumi/pulumi/issues/2484 and https://github.com/pulumi/pulumi/issues/5941.